lib/spack/llnl/util/filesystem.py

```diff
@@ -1548,3 +1550,42 @@ def search_paths_for_executables(*path_hints):
             executable_paths.append(bin_dir)
 
     return executable_paths
+
+
+def partition_path(path, entry=None):
+    """
+    Split the prefixes of the path at the first occurrence of entry and
+    return a 3-tuple containing a list of the prefixes before the entry, a
+    string of the prefix ending with the entry, and a list of the prefixes
+    after the entry.
+
+    If the entry is not a node in the path, the result will be the prefix list
+    followed by an empty string and an empty list.
+    """
+    paths = prefixes(path)
+
+    if entry is not None:
+        # Derive the index of entry within paths, which will correspond to
+        # the location of the entry in within the path.
+        try:
+            entries = path.split(os.sep)
+            i = entries.index(entry)
+            if '' in entries:
+                i -= 1
+            return paths[:i], paths[i], paths[i + 1:]
+        except ValueError:
+            pass
+
+    return paths, '', []
+
+
+def prefixes(path):
+    """Returns a list containing the path and its ancestors, top-to-bottom."""
```

Comment:
I think this could use a bit more explanation: I think it would be hard for someone to understand what this function returns by reading just the docstring. An example (added to the docstring itself) would help. Also IMO the docstring should explain that `prefixes` does not contain `/` for absolute paths.

Also, I think that while `[p or os.sep for p in path.rstrip(os.sep).split(os.sep)]` contains all the logic necessary to handle both absolute and relative paths, it is a bit much to unpack. I'd prefer something more blunt like:

```
if not path:
  return []

if path.startswith(os.sep):
  parts = [os.sep]
else:
  parts = []
path = path.lstrip(os.sep).rstrip(os.sep) # remove / from both ends
parts += list(path.split(os.sep))
```

Reply:
Changing the docstring.

The code you're complaining about was what Todd suggested at https://github.com/spack/spack/pull/12516#discussion_r320559905.

Reply:
Also, you can remove `os.sep` from path simply using `strip`:

```path = path.strip(os.sep)```

Reply:
"Simplified" as in changed use of list comprehension for `parts`